libs/algo/odc/algo/_tiff.py

```diff
@@ -0,0 +1,514 @@
+                dst.close()
+        elif idx < len(self._layers):
+            self._layers[idx].close()
+
+    def _copy_cog(self, extract=False, strict=False) -> Optional[bytes]:
+        with rasterio.Env(
+            GDAL_TIFF_OVR_BLOCKSIZE=self._ovr_blocksize,
+            GDAL_DISABLE_READDIR_ON_OPEN=False,
+            NUM_THREADS="ALL_CPUS",
+            GDAL_NUM_THREADS="ALL_CPUS",
+        ):
+            src = self._layers[0].name
+            if self._mem is not None:
+                rio_copy(
+                    src,
+                    self._mem.name,
+                    copy_src_overviews=True,
+                    strict=strict,
+                    **self._rio_opts,
+                )
+                if extract:
+                    # NOTE: this creates a copy of compressed bytes
+                    return bytes(self._mem.getbuffer())
+            else:
+                rio_copy(
+                    src,
+                    self._dst,
+                    copy_src_overviews=True,
+                    strict=strict,
+                    **self._rio_opts,
+                )
+            return None
+
+    def finalise(self, extract=False, strict=False) -> Optional[bytes]:
+        self.close()  # Write out any remainders if needed
+        return self._copy_cog(strict=strict, extract=extract)
+
+    def mem(self):
+        return self._mem
+
+    def dump_to_s3(self, url, **kw):
+        import boto3
+        from boto3.s3.transfer import TransferConfig
+        from odc.aws import s3_url_parse
+
+        assert self._mem is not None
+
+        GB = 1 << 30
+        transfer_config = TransferConfig(multipart_threshold=5 * GB)
+        bucket, key = s3_url_parse(url)
+        s3 = boto3.client("s3")
+
+        return s3.upload_fileobj(
+            self._mem, bucket, key, ExtraArgs=kw, Config=transfer_config
+        )
+
+    @staticmethod
+    def dask_finalise(sink: Delayed, extract=False, strict=False, *deps) -> Delayed:
+        """
+
+        When extract=True --> returns bytes (doubles memory requirements!!!)
+        When extract=False -> returns sink after completing everything
+        """
+        delayed_close = dask.delayed(lambda sink, idx, *deps: sink.close(idx))
+        parts = [
+            delayed_close(sink, idx, *deps, dask_key_name=("cog_close", idx))
+            for idx in range(8)
+        ]
+
+        def _copy_cog(sink, extract, strict, *parts):
+            bb = sink._copy_cog(extract=extract, strict=strict)
+            return bb if extract else sink
+
+        return dask.delayed(_copy_cog)(
+            sink, extract, strict, *parts, dask_key_name="cog"
+        )
+
+
+def save_cog(
+    xx: xr.DataArray,
+    dst: str,
+    blocksize: Optional[int] = None,
+    ovr_blocksize: Optional[int] = None,
+    bigtiff: Union[bool, str] = "auto",
+    temp_folder: Optional[str] = None,
+    overview_resampling: str = "average",
+    use_final_blocksizes: bool = False,
```

Comment:
This argument isn't being passed through to the `COGSink` constructor and should be.

Reply:
good catch fixed that and also first pass compression options